minifi/minifi-c2/minifi-c2-service/src/main/java/org/apache/nifi/minifi/c2/service/SimpleC2ProtocolService.java

```diff
@@ -0,0 +1,120 @@
+        // This service assumes there is a single Operation UPDATE to pass over the updated flow
+        Marker marker = LogMarkerUtil.getMarker(operationAck);
+
+        logger.debug(marker, "Received operation acknowledgement: {}; {}", operationAck, context);
+        // Remove the operator ID from the list of issued operations and log the state
+        final String operationId = operationAck.getOperationId();
+        try {
+            OperationState opState = OperationState.DONE;
+            String details = null;
+
+            /* Partial applications are rare and only happen when an operation consists of updating multiple config
+             * items and some succeed ( we don't yet have the concept of rollback in agents ).
+             * Fully Applied yields an operation success.
+             * Operation Not Understood and Not Applied give little details but also will result in Operation Failure.
+             * We should explore if providing textual details. */
+            final C2OperationState c2OperationState = operationAck.getOperationState();
+            if (null != c2OperationState) {
+                details = c2OperationState.getDetails();
+                if (c2OperationState.getState() != C2OperationState.OperationState.FULLY_APPLIED) {
+                    opState = OperationState.FAILED;
+                }
+            }
+
+            if (!issuedOperationIds.remove(operationId)) {
+                logger.warn(marker, "Operation with ID " + operationId + " has either already been acknowledged or is unknown to this server");
+            } else if (null != c2OperationState) {
+                final C2OperationState.OperationState operationState = c2OperationState.getState();
+                logger.debug("Operation with ID " + operationId + " acknowledged with a state of " + operationState.name() + "(" + opState.name() + "), details = "
+                        + (details == null ? "" : details));
+            }
+
+            // Optionally, an acknowledgement can include some of the info normally passed in a heartbeat.
+            // If this info is present, process it as a heartbeat, so we update our latest known state of the agent.
+            if (operationAck.getAgentInfo() != null
+                    || operationAck.getDeviceInfo() != null
+                    || operationAck.getFlowInfo() != null) {
+                final C2Heartbeat heartbeatInfo = toHeartbeat(operationAck);
+                logger.trace(marker, "Operation acknowledgement contains additional info. Processing as heartbeat: {}", heartbeatInfo);
+                processHeartbeat(heartbeatInfo, context);
+            }
+
+        } catch (final Exception e) {
+            logger.warn("Encountered exception while processing operation ack", e);
+        }
+    }
+
+    @Override
+    public C2HeartbeatResponse processHeartbeat(final C2Heartbeat heartbeat, final C2ProtocolContext context) {
```

Comment:
With this implementation at every incoming heartbeat we are sending an update operation to the agent so the agent has to download the flow after each heartbeat. Was this the intention? Maybe we would need to track that which agent downloaded (acknowledged) the flow already and issue update operation only for the agents who has not downloaded it yet.

Reply:
The intention was for it to work the way the PullHttpChangeIngestor worked so I'll check that out and make sure for now it does the same. I believe that has its own "heartbeat" rate and the flow is returned each time. We can make the "real" heartbeat stuff more sophisticated later if need be, or save that for the full C2 server implementation. Thoughts?

Reply:
I think for now what we have should be good I was just worried that each time the unchanged flow is re-downloaded unnecessarily. If the agent doesn't apply the flow if it is unchanged we should be good and I agree we should make this more robust when the full C2 implementation comes I suppose it is not in scope for this change.

Reply:
I agree with Csaba, this feels wrong to me. At the very least, it would be good to check the incoming heartbeat flow info to do some sort of comparison to determine if the update flow operation should be sent. Otherwise, the agent will need to add logic to determine "is this change necessary?"

Reply:
I'll look into doing a comparison so we don't send the same flow to the agent if it already has it.